internal/zinc-core/src/main/scala/sbt/internal/inc/Stamp.scala

```diff
@@ -58,10 +59,10 @@ trait WithPattern { protected def Pattern: Regex }
 import java.lang.{ Long => BoxedLong }
 
 /** Define the hash of the file contents. It's a typical stamp for compilation sources. */
-final class Hash(val value: Array[Byte]) extends StampBase {
-  val hexHash = IOHash.toHex(value)
+final class Hash private[inc] (val hexHash: String) extends StampBase {
```

Comment:
The validation is in the types. `Stamp` needs to pass in a string (which is the deserialized format). Adding a method in Hash's companion object does not solve anything: instead of exposing the constructor with string you will expose a method that has the same signature. People can still call that from anywhere in Zinc.